src/rtc_pl031.rs

```diff
@@ -0,0 +1,612 @@
+///
+/// This structure emulates the registers for the RTC and helps in
+/// the interaction between the driver and device by using a
+/// [`Trigger`](../trait.Trigger.html) object for notifications.
+///
+/// # Example
+///
+/// ```rust
+/// # use std::thread;
+/// # use std::io::{Error, Result};
+/// # use std::ops::Deref;
+/// # use std::time::{Instant, Duration, SystemTime, UNIX_EPOCH};
+/// # use vm_superio::Trigger;
+/// # use vm_superio::RTC;
+/// # use vmm_sys_util::eventfd::EventFd;
+///
+/// struct EventFdTrigger(EventFd);
+/// impl Trigger for EventFdTrigger {
+///     type E = Error;
+///
+///     fn trigger(&self) -> Result<()> {
+///         self.write(1)
+///     }
+/// }
+/// impl Deref for EventFdTrigger {
+///     type Target = EventFd;
+///     fn deref(&self) -> &Self::Target {
+///         &self.0
+///     }
+/// }
+/// impl EventFdTrigger {
+///     pub fn new(flag: i32) -> Self {
+///         EventFdTrigger(EventFd::new(flag).unwrap())
+///     }
+///     pub fn try_clone(&self) -> Self {
+///         EventFdTrigger((**self).try_clone().unwrap())
+///     }
+/// }
+///
+/// let mut data = [0; 4];
+/// let intr_evt = EventFdTrigger::new(libc::EFD_NONBLOCK);
+/// let mut rtc = RTC::new(intr_evt.try_clone());
+/// const RTCDR: u64 = 0x0; // Data Register.
+/// const RTCLR: u64 = 0x8; // Load Regiser.
+///
+/// // Write system time since UNIX_EPOCH in seconds to the Load Register
+/// let v = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+/// data = (v as u32).to_le_bytes();
+/// rtc.write(RTCLR, &data).unwrap();
+///
+/// // Read the value back out of the Load Register
+/// rtc.read(RTCLR, &mut data);
+/// assert_eq!((v as u32), u32::from_le_bytes(data));
+///
+/// // Sleep for 1.5 seconds to let the counter tick.
+/// let delay = Duration::from_millis(1500);
+/// thread::sleep(delay);
+///
+/// // Read the current RTC value from the Data Register
+/// rtc.read(RTCDR, &mut data);
+/// assert!(u32::from_le_bytes(data) > (v as u32));
+/// ```
+pub struct RTC<T: Trigger> {
+    // Counts up from 1 on reset at 1Hz (emulated).
+    counter: Instant,
+
+    // The offset value applied to the counter to get the RTC value.
+    counter_offset: u32,
+
+    // The comparison value used as an alarm (not implemented).
+    // TODO: Implement the match register functionality.
+    // match_value: u32,
+
+    // The interrupt mask.
+    imsc: u32,
+
+    // The raw interrupt value.
+    ris: u32,
+
+    // Used for notifying the driver about events.
+    interrupt_evt: T,
+}
+
+/// Errors encountered while handling RTC operations.
+#[derive(Debug)]
+pub enum Error<E> {
+    /// Failed to trigger interrupt.
+    Trigger(E),
+    /// Attempting to write to a non-existent register offset
+    /// or to the match register, which is not emulated.
+    BadWriteOffset(u64),
+    /// Attempting to write to a non-existent register offset
+    /// or to the match register, which is not emulated.
+    BadReadOffset(u64),
```

Comment:
The `BadReadOffset` error does not appear to be used; `RTC::read()` silently discards invalid reads (as in the original Firecracker code). Perhaps this error should be removed?

Reply:
I agree.  Removed.